feat!: Insert fewer order edges when inlining and update spec for order edges (#3058)
Quantinuum/hugr: specification/hugr.md

### `Order` edges

> `Order` edges provide a way to express constraints on ordering between nodes

Author: I wonder if "express constraints on ordering" should be "to express hidden/stateful dependences" - they no longer mean the first node has to finish before the second node starts, unless the nodes are atomic....thoughts?

    linkStyle 12,13,14,15,16,17 stroke:#ff3,stroke-width:4px;
```

(evaluation-semantics)=
## Evaluation Semantics

> In a dataflow graph, the evaluation semantics are simple: all nodes in the

Author: This moved from its previous location which seemed a bit out-of-place, and significantly expanded, but I kept the original definition of all nodes being evaluated; I am tempted to say "at least those nodes required to produce all inputs to the Output node must be evaluated; more may be at the discretion of the implementation", although this is really targetted at (a) Tierkreis (b) removing nonterminating computations whose results are unused; thoughts?

Moreover, our current implementation of dead-code elimination transforms the hugr to another that *does not conform* to these evaluation semantics (removed nodes are not evaluated!); should we say something like "evaluating a dataflow graph behaves as if all nodes were evaluated"?

Reviewer: Yes, if we say something like "evaluating a dataflow graph behaves as if all nodes were evaluated", this covers all bases I think?

graph are necessarily evaluated. (For hierarchical nodes, this may entail
evaluating some children, according to the node.) Evaluation may occur in
some order, and for nodes that execute strictly and atomically (according to

> implementation) this will respect the Dataflow and Order edges. However *in

Reviewer: ```suggestion
evaluating some children, according to the node.) For nodes that execute strictly and atomically (according to
implementation) the order of execution will respect the Dataflow and Order edges. However *in
```
Is this clearer, or am I missing a subtlety?

    linkStyle 12,13,14,15,16,17 stroke:#ff3,stroke-width:4px;
```

(evaluation-semantics)=
## Evaluation Semantics

In a dataflow graph, the evaluation semantics are simple: all nodes in the
graph are necessarily evaluated. (For hierarchical nodes, this may entail
evaluating some children, according to the node.) For nodes that execute
strictly (according to implementation) this will happen in an order that

> respects the Dataflow and Order edges; this will be a total order if the

Reviewer: Can you clarify what this means? In particular, "For nodes that execute strictly (according to implemention)".

Reviewer: (See also https://github.com/Quantinuum/tket2/discussions/1649 .)

Author: There were two things I was trying to get at:
* For nodes that execute strictly, they execute in an order respecting the edges; it's a total order if they execute atomically (i.e. no overlapping).
   * ....I guess if non-atomic then it might be a partial order, but then, if nonstrict you could say a partial order too, and both the non-total cases are stretching the meaning of a "partial order" - the partial ordering where nothing is before or after anything else is still partial but pretty useless - so I didn't mention that
* Both strictness and also atomicity depend on the specific toolchain or implementation

In the end I expanded, rephrased, deleted some repetition, and added a footnote clarifying "strict", so this did change a bit (again), but only in presentation not meaning, promise ;-)

Author: Ok removed the business about total order, it's not really helpful, and may just confuse. The key point is that without strictness, evaluation order doesn't even need to respect the edges.

nodes execute atomically. However *in general* Hugr allows that
* Nodes may execute non-atomically, perhaps in parallel/overlapping other nodes
* Nodes may execute partially or totally before all inputs are ready (if the

>   op is able to "do" anything without all its inputs). **However** note that

Reviewer: ```suggestion
  op is able to "do" anything without all its inputs). But note that
```
Just a style thing (avoiding nested "howevers").

Author: I did a bit more bracket removal and cleanup around here and ended up with this being the only However. Kept the emphasis though as it's not a consequence but its own rule.

  a node $n$ with an `Order` input must wait for all its `Order`-predecessors
  to produce their `Order` output before:
    * $n$ has any side-effects (e.g. panic or print)
    * producing $n$'s `Order` output
  this means that side-effects will occur in a partial/total order respecting
  the `Order` edges, but the `Order` edges do not restrict pure functional
  computation (that does not depend on any global state or have any side effects).

This applies both to extension ops and core constructs such as DFG, CFG, or a
Call to a function, but may be refined (introducing more precise guarantees
such as atomicity or strictness) for specific ops and/or by specific
implementations.

> For ops such as `Conditional`, `TailLoop` and `CFG`/`BasicBlock`, the order input to the container node, or to any `Output` node within, must be treated linearly: (only) after all Order-predecessors have (finished mutating state/produced their Order outputs), may the Order output (of exactly one `Input` node or the container itself) be ready - chosen according to the appropriate predicate. **Thus**, nodes reachable along Order edges from an `Input` may not be speculatively evaluated (perform any stateful update); but nodes which are not `Order`-reachable may be.

Reviewer: ```suggestion
For ops such as `Conditional`, `TailLoop` and `CFG`/`BasicBlock`, the order input to the container node, or to any `Output` node within, must be treated linearly: only after all Order-predecessors have finished mutating state and produced their Order outputs may the Order output (of exactly one `Input` node or the container itself) be ready - chosen according to the appropriate predicate. Thus, nodes reachable along Order edges from an `Input` may not be speculatively evaluated (perform any stateful update); but nodes which are not Order-reachable may be.
```

Author: Yes, much better, thanks
